File: 2026-01-January-GFG-POTD/January 21 - Stock span problem/Solution.java
class Solution {
    public ArrayList<Integer> calculateSpan(int[] arr) {
        ArrayList<Integer> span = new ArrayList<>();
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < arr.length; i++) {

            // Pop while current price is higher or equal
            while (!stack.isEmpty() && arr[stack.peek()] <= arr[i]) {
                stack.pop();
            }

            // If stack is empty, span is i + 1
            int currSpan = stack.isEmpty() ? i + 1 : i - stack.peek();
            span.add(currSpan);

            // Push current index
            stack.push(i);
        }

        return span;
    }
}

File: 2026-01-January-GFG-POTD/January 21 - Stock span problem/Solution.py
class Solution:
    def calculateSpan(self, arr):
        span = []
        stack = []

        for i in range(len(arr)):

            # Pop while current price is higher or equal
            while stack and arr[stack[-1]] <= arr[i]:
                stack.pop()

            curr_span = i + 1 if not stack else i - stack[-1]
            span.append(curr_span)

            # Push current index
            stack.append(i)

        return span
